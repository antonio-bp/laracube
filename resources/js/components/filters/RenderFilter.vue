<template>
    <div v-if="!fetching">
        <div v-if="filter.type === 'single-select'">
            <single-select
                :report="report"
                :filter="filter"
                :items="response"
                :store-selected="getStoreSelected()"
                @filter-changed="filterChanged"
            ></single-select>
        </div>
        <div v-if="filter.type === 'multiple-select'">
            <multiple-select :filter="filter" :report="report" :response="response"></multiple-select>
        </div>
    </div>
</template>

<script>
import SingleSelect from '@/components/filters/elements/SingleSelect';
import MultipleSelect from '@/components/filters/elements/MultipleSelect';
import { mapState } from 'vuex';
export default {
    name: 'RenderFilter',
    components: { MultipleSelect, SingleSelect },
    props: {
        report: { required: true },
        filter: { required: true },
    },
    computed: {
        url() {
            return '/laracube-api/run/filter/' + this.filter.uriKey;
        },
        ...mapState('filters', ['filters']),
    },
    data() {
        return {
            fetching: true,
            response: {},
        };
    },
    mounted() {
        this.fetchData();
    },
    methods: {
        fetchData() {
            this.fetching = true;
            if (this.url === undefined) {
                throw 'url data property not defined';
            }
            this.$axios
                .post(this.url)
                .then(({ data }) => {
                    this.response = data;
                })
                .catch((error) => {
                    console.log(error);
                    this.$toasted.error(error.message);
                })
                .finally(() => {
                    this.fetching = false;
                });
        },
        getStoreSelected() {
            // check if there is a selected value for this filter
            if (
                this.filters.hasOwnProperty(this.report.meta.uriKey) &&
                this.filters[this.report.meta.uriKey].hasOwnProperty(this.filter.key)
            ) {
                return this.filters[this.report.meta.uriKey][this.filter.key];
            }
            return null;
        },
        filterChanged(args) {
            this.$emit('filter-changed', args);
        },
    },
};
</script>
