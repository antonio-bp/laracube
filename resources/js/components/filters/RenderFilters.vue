<template>
    <div>
        <div v-if="Array.isArray(report.filters) && report.filters.length">
            <v-dialog v-model="filterDialog" fullscreen hide-overlay transition="dialog-bottom-transition">
                <template #activator="{ on: dialog }">
                    <v-tooltip left>
                        <template #activator="{ on: tooltip }">
                            <v-badge :content="appliedFilterCount" :value="appliedFilterCount" color="primary">
                                <v-icon v-on="{ ...tooltip, ...dialog }"> mdi-filter-variant </v-icon>
                            </v-badge>
                        </template>
                        <span>Filters</span>
                    </v-tooltip>
                </template>
                <v-card>
                    <v-toolbar dark color="primary" tile flat class="lc-shadow">
                        <v-btn icon dark @click="closeDialog">
                            <v-icon>mdi-close</v-icon>
                        </v-btn>
                        <v-toolbar-title>Filters</v-toolbar-title>
                        <v-spacer></v-spacer>
                        <v-toolbar-items>
                            <v-btn dark text @click="resetAll"> <v-icon>mdi-refresh</v-icon> Reset all </v-btn>
                        </v-toolbar-items>
                    </v-toolbar>
                    <v-list three-line subheader class="mt-5">
                        <v-list-item v-for="filter in report.filters" :key="filter.key">
                            <v-list-item-content>
                                <v-list-item-title>
                                    {{ filter.heading }}
                                </v-list-item-title>
                                <div v-if="filter.type === 'single-select'">
                                    <single-select
                                        :report="report"
                                        :filter="filter"
                                        :refresh="filterDialog"
                                        @filter-changed="filterChanged"
                                        :key="filter.key"
                                    ></single-select>
                                </div>
                                <div v-if="filter.type === 'multiple-select'">
                                    <multiple-select
                                        :report="report"
                                        :filter="filter"
                                        :refresh="filterDialog"
                                        @filter-changed="filterChanged"
                                        :key="filter.key"
                                    ></multiple-select>
                                </div>
                            </v-list-item-content>
                        </v-list-item>
                        <v-list-item>
                            <v-list-item-content>
                                <v-btn
                                    elevation="0"
                                    tile
                                    large
                                    color="primary"
                                    class="lc-shadow px-15"
                                    @click="applyFilters"
                                >
                                    Apply Filters
                                </v-btn>
                            </v-list-item-content>
                        </v-list-item>
                    </v-list>
                </v-card>
            </v-dialog>
        </div>
    </div>
</template>

<script>
import { mapMutations } from 'vuex';
import SingleSelect from '@/components/filters/elements/SingleSelect';
import MultipleSelect from '@/components/filters/elements/MultipleSelect';
import _ from 'lodash';

export default {
    name: 'RenderFilters',
    components: { SingleSelect, MultipleSelect },
    props: {
        report: { required: true },
    },
    mounted() {
        if (Array.isArray(this.report.filters) && this.report.filters.length) {
            this.resetFilters();
            if (!this.reportFilters.hasOwnProperty(this.report.meta.uriKey)) {
                this.reportFilters[this.report.meta.uriKey] = {};
            }
        }
    },
    computed: {
        appliedFilterCount() {
            if (this.reportFilters.hasOwnProperty(this.report.meta.uriKey)) {
                return _.size(this.reportFilters[this.report.meta.uriKey]);
            }
            return 0;
        },
    },
    data() {
        return {
            filterDialog: null,
            reportFilters: {},
        };
    },
    methods: {
        ...mapMutations('filters', {
            setFilters: 'SET_FILTERS',
        }),
        filterChanged(args) {
            let condition =
                !!(!Array.isArray(args.selected) && args.selected) ||
                !!(Array.isArray(args.selected) && args.selected.length);
            if (condition) {
                this.reportFilters[this.report.meta.uriKey][args.filterKey] = args.selected;
            } else {
                delete this.reportFilters[this.report.meta.uriKey][args.filterKey];
            }
        },
        applyFilters() {
            this.setFilters(_.cloneDeep(this.reportFilters));
            this.$emit('filter-applied');
            this.resetFilters();
            this.filterDialog = false;
        },
        closeDialog() {
            this.resetFilters();
            this.filterDialog = false;
        },
        resetAll() {
            this.reportFilters[this.report.meta.uriKey] = {};
            this.setFilters(_.cloneDeep(this.reportFilters));
            this.$emit('filter-applied');
            this.resetFilters();
            this.filterDialog = false;
        },
        resetFilters() {
            this.reportFilters = _.cloneDeep(this.$store.state.filters.filters);
        },
    },
};
</script>
