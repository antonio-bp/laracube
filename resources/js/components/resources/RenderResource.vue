<template>
    <div class="fill-height">
        <div v-if="resource.component === 'table'" class="fill-height">
            <laracube-table :resource="resource"></laracube-table>
        </div>
        <div v-if="resource.component === 'card'" class="fill-height">
            <card :resource="resource"></card>
        </div>
    </div>
</template>

<script>
import LaracubeTable from '@/components/resources/Table';
import Card from '@/components/resources/Card';

export default {
    name: 'RenderResource',
    components: { Card, LaracubeTable },
    props: {
        resource: { required: true },
    },
};
</script>
