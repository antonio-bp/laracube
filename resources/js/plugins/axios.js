import ax from 'axios';
import store from '@/store';

ax.interceptors.request.use(
    function (config) {
        let source = ax.CancelToken.source();
        config.cancelToken = source.token;
        store.commit('axios/ADD_CANCEL_TOKEN', source);
        return config;
    },
    function (error) {
        return Promise.reject(error);
    },
);

export const axios = ax;

export default {
    install(Vue, options) {
        Vue.prototype.$axios = ax;
    },
};
