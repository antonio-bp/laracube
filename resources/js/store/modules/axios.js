export default {
    namespaced: true,
    state: {
        cancelTokens: [],
    },
    getters: {
        cancelTokens(state) {
            return state.cancelTokens;
        },
    },
    mutations: {
        ADD_CANCEL_TOKEN(state, token) {
            state.cancelTokens.push(token);
        },
        CLEAR_CANCEL_TOKENS(state) {
            state.cancelTokens = [];
        },
    },
    actions: {
        CANCEL_PENDING_REQUESTS(context) {
            context.state.cancelTokens.forEach((request, i) => {
                if (request.cancel) {
                    request.cancel();
                }
            });
            context.commit('CLEAR_CANCEL_TOKENS');
        },
    },
};
