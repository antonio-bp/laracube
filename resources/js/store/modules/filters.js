export default {
    namespaced: true,
    state: {
        filters: {},
    },
    mutations: {
        SET_FILTERS(state, data) {
            state.filters = data;
        },
    },
};
