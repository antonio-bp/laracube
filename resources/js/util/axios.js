import store from '@/store';

window.axios = require('axios');

window.axios.defaults.headers.common['X-Requested-With'] = 'XMLHttpRequest';

window.axios.interceptors.request.use(
    function (config) {
        let source = window.axios.CancelToken.source();
        config.cancelToken = source.token;
        store.commit('axios/ADD_CANCEL_TOKEN', source);
        return config;
    },
    function (error) {
        return Promise.reject(error);
    },
);
