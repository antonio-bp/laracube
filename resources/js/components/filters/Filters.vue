<template>
    <div>
        <div v-if="Array.isArray(report.filters) && report.filters.length">
            <v-dialog v-model="filterDialog" fullscreen hide-overlay transition="dialog-bottom-transition">
                <template #activator="{ on: dialog }">
                    <v-tooltip left>
                        <template #activator="{ on: tooltip }">
                            <v-icon v-on="{ ...tooltip, ...dialog }"> mdi-filter-variant </v-icon>
                        </template>
                        <span>Filters</span>
                    </v-tooltip>
                </template>
                <v-card>
                    <v-toolbar dark color="primary" tile flat class="lc-shadow">
                        <v-btn icon dark @click="filterDialog = false">
                            <v-icon>mdi-close</v-icon>
                        </v-btn>
                        <v-toolbar-title>Filters</v-toolbar-title>
                        <v-spacer></v-spacer>
                        <v-toolbar-items>
                            <v-btn dark text @click="resetFilters"> <v-icon>mdi-refresh</v-icon> Reset all </v-btn>
                        </v-toolbar-items>
                    </v-toolbar>
                    <v-list three-line subheader class="mt-5">
                        <v-list-item v-for="filter in report.filters" :key="filter.key">
                            <v-list-item-content>
                                <v-list-item-title>
                                    {{ filter.heading }}
                                </v-list-item-title>
                                <render-filter
                                    :report="report"
                                    :filter="filter"
                                    @filter-changed="filterChanged"
                                ></render-filter>
                            </v-list-item-content>
                        </v-list-item>
                        <v-list-item>
                            <v-list-item-content>
                                <v-btn
                                    elevation="0"
                                    tile
                                    large
                                    color="primary"
                                    class="lc-shadow px-15"
                                    @click="applyFilters"
                                >
                                    Apply Filters
                                </v-btn>
                            </v-list-item-content>
                        </v-list-item>
                    </v-list>
                </v-card>
            </v-dialog>
        </div>
    </div>
</template>

<script>
import RenderFilter from '@/components/filters/RenderFilter';
import { mapMutations, mapState } from 'vuex';

export default {
    name: 'Filters',
    components: { RenderFilter },
    props: {
        report: { required: true },
    },
    computed: {
        ...mapState('filters', ['filters']),
    },
    mounted() {
        if (Array.isArray(this.report.filters) && this.report.filters.length) {
            this.reportFilters = { ...this.filters };
            // set report filters
            if (!this.reportFilters.hasOwnProperty(this.report.meta.uriKey)) {
                this.reportFilters[this.report.meta.uriKey] = {};
            }
            for (const filter in this.report.filters) {
                if (!this.reportFilters.hasOwnProperty(this.report.meta.uriKey)) {
                    this.reportFilters[this.report.meta.uriKey] = {};
                }
                console.log(this.report.filters[filter]);
            }
            // let reportFilters = this.reportFilters[this.report.meta.uriKey];
            //
            // for (i = 0; i < this.report.filters.length; i++) {
            //     if (
            //         !this.reportFilters[this.report.meta.uriKey].hasOwnProperty(
            //             this.report.filter[i].key,
            //         )
            //     ) {
            //         this.reportFilters[this.report.meta.uriKey] = {};
            //     }
            //
            //     this.report.filters.forEach(function (filter) {
            //         if (!reportFilters.hasOwnProperty(filter.key)) {
            //             this.reportFilters[this.report.meta.uriKey][
            //                 filter.key
            //             ] = {};
            //         }
            //     });

            // for (filter in ) {
            //     if (!reportFilters.hasOwnProperty(filter.key)) {
            //         this.reportFilters[this.report.meta.uriKey] = {};
            //     }
            // }
        }
    },
    data() {
        return {
            filterDialog: null,
            reportFilters: {},
        };
    },
    methods: {
        ...mapMutations('filters', {
            setFilters: 'SET_FILTERS',
        }),
        filterChanged(args) {
            // set the new local report filter
            console.log('set the new local report filter');
            console.log(args.filterKey);
            console.log(args.selected);
            this.reportFilters[this.report.meta.uriKey][args.filterKey] = args.selected;
        },
        applyFilters() {
            // save the filter in store
            this.setFilters(this.reportFilters);
            this.$emit('filter-applied');
            this.filterDialog = false;
        },
        resetFilters() {
            this.reportFilters[this.report.meta.uriKey] = {};
            this.setFilters(this.reportFilters);
            this.$emit('filter-applied');
            this.filterDialog = false;
        },
    },
};
</script>
