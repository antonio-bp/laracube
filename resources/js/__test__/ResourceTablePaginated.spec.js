import { createLocalVue, mount } from '@vue/test-utils';
import Vue from 'vue';
import Vuex from 'vuex';
import Vuetify from 'vuetify';
import RenderResource from '@/components/resources/RenderResource';
import flushPromises from 'flush-promises';
import axios from 'axios';
import filters from '@/store/modules/filters';

Vue.use(Vuetify);
const localVue = createLocalVue();
localVue.use(Vuex);

jest.mock('axios', () => ({
    post: () =>
        Promise.resolve({
            data: {
                data: [{ name: 'Marvin' }, { name: 'Jasper' }],
                meta: {
                    last_page: 5,
                    from: 1,
                    to: 3,
                    total: 20,
                    columns: [
                        {
                            sortable: false,
                            text: 'Name',
                            tooltip: null,
                            value: 'name',
                        },
                    ],
                },
            },
        }),
}));

function factory() {
    let vuetify = new Vuetify();
    let store = new Vuex.Store({
        modules: {
            filters: {
                state: filters.state,
                mutations: filters.mutations,
                namespaced: true,
            },
        },
    });

    return mount(RenderResource, {
        store,
        localVue,
        vuetify,
        propsData: {
            resource: {
                columns: 4,
                component: 'table',
                heading: 'Heading',
                subHeading: 'Sub Heading',
                type: 'paginated',
                uriKey: 'paginated-table',
            },
            report: {
                meta: {
                    uriKey: 'net-revenue-report',
                },
            },
        },
        mocks: {
            $axios: axios,
        },
    });
}

describe('TablePaginated', () => {
    it('should load the components', () => {
        const wrapper = factory();
        expect(wrapper.find('.v-card').vm).toBeTruthy();
    });

    it('should show the heading and sub-heading tooltip', () => {
        const wrapper = factory();
        expect(wrapper.find('.v-card__title').html()).toContain('Heading');
        expect(wrapper.find('.v-tooltip').vm).toBeTruthy();
    });

    it('should show the correct data from api response', async () => {
        const wrapper = factory();
        await flushPromises();
        expect(wrapper.html()).toMatchSnapshot();
        expect(wrapper.find('td').html()).toContain('Marvin');
    });
});
