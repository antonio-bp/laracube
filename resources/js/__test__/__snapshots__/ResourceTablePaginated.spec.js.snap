// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TablePaginated should show the correct data from api response 1`] = `
<div class="fill-height">
  <div class="fill-height">
    <div class="lc-shadow lc-rounded fill-height v-card v-card--flat v-sheet theme--light">
      <div class="v-card__title grey--text text--darken-2 pr-2">Heading</div> <span class="v-tooltip v-tooltip--right"><!----></span><i aria-haspopup="true" aria-expanded="false" class="fas fa-info-circle grey--text text--darken-1"></i>
      <div class="v-card__text px-0">
        <div>
          <div class="v-data-table theme--light">
            <div class="v-data-table__wrapper">
              <table>
                <colgroup>
                  <col class="">
                </colgroup>
                <thead class="v-data-table-header">
                  <tr>
                    <th role="columnheader" scope="col" aria-label="Name" class="text-start"><span class="grey--text text--darken-2">Name</span></th>
                  </tr>
                </thead>
                <tbody>
                  <tr class="">
                    <td class="text-start">
                      <div>Marvin</div>
                    </td>
                  </tr>
                  <tr class="">
                    <td class="text-start">
                      <div>Jasper</div>
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
          </div>
          <div class="row pt-2 align-center no-gutters">
            <div class="pl-5 grey--text text--darken-2 col">
              Showing
              <span class="font-weight-bold">
                        1
                    </span>
              to
              <span class="font-weight-bold">
                        3
                    </span>
              of
              <span class="font-weight-bold">
                        20
                    </span>
              results
            </div>
            <div class="col">
              <nav role="navigation" aria-label="Pagination Navigation" class="float-right pr-3 grey--text text--darken-2">
                <ul class="v-pagination theme--light">
                  <li><button disabled="disabled" type="button" aria-label="Previous page" class="v-pagination__navigation v-pagination__navigation--disabled"><i aria-hidden="true" class="v-icon notranslate mdi mdi-chevron-left theme--light"></i></button></li>
                  <li><button type="button" aria-current="true" aria-label="Current Page, Page 1" class="v-pagination__item v-pagination__item--active primary">1</button></li>
                  <li><button type="button" aria-label="Goto Page 2" class="v-pagination__item">2</button></li>
                  <li><button type="button" aria-label="Goto Page 3" class="v-pagination__item">3</button></li>
                  <li><button type="button" aria-label="Goto Page 4" class="v-pagination__item">4</button></li>
                  <li><button type="button" aria-label="Goto Page 5" class="v-pagination__item">5</button></li>
                  <li><button type="button" aria-label="Next page" class="v-pagination__navigation"><i aria-hidden="true" class="v-icon notranslate mdi mdi-chevron-right theme--light"></i></button></li>
                </ul>
              </nav>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <!---->
</div>
`;
